### PR TITLE
clean work tree folder before using

### PR DESCRIPTION
**Pull request by OsmondJiang**
#4542 

This PR is add more checking to the test, make sure we know the root cause if it happened again.
And also found some business logic bug. The newly added checks will cover these part changes

### CODE REVIEW

File: src/docfx/restore/index/DependencySlotPool.cs [hunk 1]
```diff
@@ -111,7 +111,7 @@ public static (string path, T slot) AcquireSlot(string url, LockType type, Func<
                             var existed = false;
                             foreach (var i in slots)
                             {
-                                if (DateTime.UtcNow - i.LastAccessDate > TimeSpan.FromSeconds(_defaultLockdownTimeInSecond))
+                                if (!i.Restored || DateTime.UtcNow - i.LastAccessDate > TimeSpan.FromSeconds(_defaultLockdownTimeInSecond))
```
Comment (mingwli) [inline src/docfx/restore/index/DependencySlotPool.cs]: >!i.Restored [](start = 36, length = 11)

the slot maybe right being used by another docfx-restore?

File: src/docfx/restore/RestoreGit.cs [hunk 2]
```diff
@@ -143,6 +143,9 @@ void AddWorkTrees()
                                         // it may be dirty caused by last failed restore action
                                         if (Directory.Exists(workTreePath))
                                             Directory.Delete(workTreePath, true);
+
+                                        // https://stackoverflow.com/questions/24265481/after-directory-delete-the-directory-exists-returning-true-sometimes
+                                        Debug.Assert(!Directory.Exists(workTreePath));
```
Comment (live1206) [inline src/docfx/restore/RestoreGit.cs]: This only fails on debug, shall we log an exception for the normal run as well?

File: src/docfx/restore/index/DependencySlotPool.cs (hunk 1 above)
Comment (OsmondJiang) [reply src/docfx/restore/index/DependencySlotPool.cs]: no, one index file can be only access by one consumer at the same time

File: src/docfx/restore/RestoreGit.cs (hunk 2 above)
Comment (OsmondJiang) [reply src/docfx/restore/RestoreGit.cs]: this is only to identify the root cause, once we know it's the root cause, we will something here.

**Pull request merged**